state-chain/pallets/cf-asset-balances/src/lib.rs

```diff
@@ -0,0 +1,312 @@
+		/// an runtime upgrade.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Handles egress for all chains.
+		type EgressHandler: EgressApi<AnyChain>;
+
+		/// Polkadot environment.
+		type PolkadotKeyProvider: KeyProvider<PolkadotCrypto>;
+
+		/// Safe mode configuration.
+		type SafeMode: Get<PalletSafeMode>;
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Refund scheduled for a validator.
+		RefundScheduled {
+			egress_id: EgressId,
+			destination: ForeignChainAddress,
+			amount: AssetAmount,
+		},
+		/// The Vault is running a deficit: we owe more than we have set aside for refunds.
+		VaultDeficitDetected {
+			chain: ForeignChain,
+			amount_owed: AssetAmount,
+			available: AssetAmount,
+		},
+	}
+
+	#[pallet::pallet]
+	#[pallet::storage_version(PALLET_VERSION)]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	/// Liabilities are funds that are owed to some external party.
+	#[pallet::storage]
+	pub type Liabilities<T: Config> =
+		StorageMap<_, Twox64Concat, Asset, BTreeMap<ExternalOwner, AssetAmount>, ValueQuery>;
+
+	/// Funds that have been set aside to refund external [Liabilities].
+	#[pallet::storage]
+	pub type WithheldAssets<T: Config> =
+		StorageMap<_, Twox64Concat, Asset, AssetAmount, ValueQuery>;
+}
+
+impl<T: Config> Pallet<T> {
+	fn refund_via_egress(
+		chain: ForeignChain,
+		address: ForeignChainAddress,
+		amount: AssetAmount,
+	) -> Result<(), DispatchError> {
+		let egress_details =
+			T::EgressHandler::schedule_egress(chain.gas_asset(), amount, address.clone(), None)
+				.map_err(Into::into)?;
+		Self::deposit_event(Event::RefundScheduled {
+			egress_id: egress_details.egress_id,
+			destination: address,
+			amount,
+		});
+		Ok(())
+	}
+
+	/// Reconciles the amount owed with the amount available for distribution.
+	///
+	/// The owed and available amount are mutated in place.
+	///
+	/// For Ethereum and Arbitrum, we expect to the validators and pay out via egress to their
+	/// accounts. For Polkadot, we expect to pay out to the current AggKey account.
+	/// For Bitcoin and Solana, the vault pays the fees directly so we don't need to egress
+	/// anything.
+	///
+	/// Note that we refund to accounts atomically (we never partially refund an account), whereas
+	/// refunds to vaults or aggkeys can be made incrementally.
+	fn reconcile(
+		chain: ForeignChain,
+		owner: &ExternalOwner,
+		amount_owed: &mut AssetAmount,
+		available: &mut AssetAmount,
+	) -> Result<(), DispatchError> {
+		if *amount_owed > *available {
+			Self::deposit_event(Event::VaultDeficitDetected {
+				chain,
+				amount_owed: *amount_owed,
+				available: *available,
+			});
+		}
+		let amount_reconciled = match chain {
+			ForeignChain::Ethereum | ForeignChain::Arbitrum => match owner {
+				ExternalOwner::Account(address) =>
+					if *amount_owed > *available {
```

Comment:
we've already checked this above

Reply:
And we need to somehow know it here again. The only solution would be is remembering the result of the condition and then using this instead, but I don't really see big benefit.

Reply:
Why do we need to know it here again? If this condition were true, we'd have already thrown an event above?

Reply:
Because we want to fire this event for any chain under this condition. For the same condition we decide if we want to send funds but only for specific chain (ETH, ARB). We need to know this because of two different reasons at two different areas of the code.